tests/tcp_listener.rs

```diff
@@ -0,0 +1,310 @@
+use std::io::Read;
+use std::net::{self, SocketAddr};
+#[cfg(unix)]
+use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd};
+use std::sync::{Arc, Barrier};
+use std::thread;
+
+use mio::net::TcpListener;
+use mio::{Interests, Poll, Token};
+
+mod util;
+
+use util::{
+    any_local_address, any_local_ipv6_address, assert_send, assert_sync, assert_would_block,
+    expect_events, expect_no_events, init, init_with_poll, ExpectEvent,
+};
+
+const ID1: Token = Token(0);
+const ID2: Token = Token(1);
+
+#[test]
+fn is_send_and_sync() {
+    assert_send::<TcpListener>();
+    assert_sync::<TcpListener>();
+}
+
+#[test]
+fn tcp_listener() {
+    smoke_test_tcp_listener(any_local_address());
+}
+
+#[test]
+fn tcp_listener_ipv6() {
+    smoke_test_tcp_listener(any_local_ipv6_address());
+}
+
+fn smoke_test_tcp_listener(addr: SocketAddr) {
+    let (mut poll, mut events) = init_with_poll();
+
+    let listener = TcpListener::bind(addr).unwrap();
+    let address = listener.local_addr().unwrap();
+
+    poll.registry()
+        .register(&listener, ID1, Interests::READABLE)
+        .expect("unable to register TCP listener");
+
+    let barrier = Arc::new(Barrier::new(2));
+    let thread_handle = start_connections(address, 1, barrier.clone());
+
+    expect_events(
+        &mut poll,
+        &mut events,
+        vec![ExpectEvent::new(ID1, Interests::READABLE)],
+    );
+
+    // Expect a single connection.
+    let (mut stream, peer_address) = listener.accept().expect("unable to accept connection");
+    assert!(peer_address.ip().is_loopback());
+    assert_eq!(stream.peer_addr().unwrap(), peer_address);
+    assert_eq!(stream.local_addr().unwrap(), address);
+
+    // Expect the stream to be non-blocking.
+    let mut buf = [0; 20];
+    assert_would_block(stream.read(&mut buf));
```

Comment:
@carllerche here's a test that ensure the stream to be non-blocking, per what you mentioned on Gitter.